File: sentil-java/src/main/java/io/github/sedislab/sentil/Controller.java
package io.github.sedislab.sentil;

/** A receding-horizon controller that emits a control input within a hard deadline. */
public final class Controller extends NativeResource {
    private final long inputWidth;
    private final long modelBox;

    private Controller(long handle, long modelBox, long inputWidth) {
        super(handle, h -> {
            NativeLib.controllerDestroy(h);
            if (modelBox != 0L) {
                NativeLib.freeModelBox(modelBox);
            }
        });
        this.inputWidth = inputWidth;
        this.modelBox = modelBox;
    }

    /**
     * A controller over the model and spec, both consumed, with budgetNs the per-step
     * deadline in nanoseconds.
     */
    public Controller(SystemModel model, Formula spec, long inputWidth, long budgetNs)
            throws SentilException {
        this(model, spec, inputWidth, budgetNs, null, null);
    }

    /** A controller with optional bounds and smoothing on the planned input. */
    public Controller(SystemModel model, Formula spec, long inputWidth, long budgetNs, Bounds bounds,
            SmoothConfig smooth) throws SentilException {
        this(build(model, spec, inputWidth, budgetNs, bounds, smooth), model.modelBox(), inputWidth);
    }

    private static long build(SystemModel model, Formula spec, long inputWidth, long budgetNs,
            Bounds bounds, SmoothConfig smooth) throws SentilException {
        long modelHandle = model.handle();
        long specHandle = spec.handle();
        long box = model.modelBox();
        model.disown();
        spec.disown();
        long boundsHandle = bounds == null ? 0L : bounds.handle();
        double temperature = smooth == null ? 0.0 : smooth.temperature;
        int kind = smooth == null ? 0 : smooth.kind.code();
        try {
            return NativeLib.controllerCreate(modelHandle, specHandle, inputWidth, budgetNs,
                    boundsHandle, temperature, kind, smooth != null);
        } catch (SentilException error) {
            if (box != 0L) {
                NativeLib.freeModelBox(box);
            }
            throw error;
        }
    }

    /** Plan from the current state and return the first control input. */
    public double[] control(double[] state) throws SentilException {
        double[] input = NativeLib.controllerControl(handle(), state, inputWidth);
        if (modelBox != 0L) {
            NativeLib.rethrowModelError(modelBox);
        }
        return input;
    }
}

File: sentil-matlab/+sentil/Controller.m
classdef Controller < handle
    %CONTROLLER A receding-horizon controller with a per-step wall-clock budget.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
        InputWidth double = 0
        ModelBox uint64 = uint64(0)
    end

    methods
        function obj = Controller(model, spec, inputWidth, budgetNs, bounds, smooth)
            %CONTROLLER Build over a model and spec, both consumed. budgetNs is in nanoseconds.
            if nargin < 5 || isempty(bounds)
                boundsHandle = uint64(0);
            else
                boundsHandle = bounds.Handle;
            end
            if nargin < 6, smooth = sentil.SmoothConfig; end
            % The C ABI consumes the model handle even on a null return.
            [modelHandle, box] = model.releaseToController();
            obj.ModelBox = box;
            obj.Handle = sentil_mex('controller_create', modelHandle, spec.consume(), ...
                double(inputWidth), double(budgetNs), boundsHandle, smooth.temperature, ...
                double(int32(smooth.kind)));
            obj.InputWidth = double(inputWidth);
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('controller_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
            if obj.ModelBox ~= 0
                sentil_mex('free_model_box', obj.ModelBox);
                obj.ModelBox = uint64(0);
            end
        end

        function u = control(obj, state)
            %CONTROL Plan from the current state and return the control input.
            obj.assertOpen();
            u = sentil_mex('controller_control', obj.Handle, double(state), obj.InputWidth, ...
                obj.ModelBox);
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this controller has been closed');
            end
        end
    end
end